### PR TITLE
New topic for Kubernetes descheduler

### PR DESCRIPTION
**Pull request by mburke5678**
https://trello.com/c/nulaT5Wp/670-document-implement-basic-descheduler

### CODE REVIEW

File: admin_guide/scheduling/descheduler.adoc [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+For more information on Red Hat Technology Preview features support scope, see
+https://access.redhat.com/support/offerings/techpreview/.
+endif::[]
+====
+
+The descheduler does not evict the following types of pods:
+
+* Critical pods (with the `scheduler.alpha.kubernetes.io/critical-pod` annotation).
+* Pods (link:https://kubernetes.io/docs/tasks/administer-cluster/static-pod/[static and mirror pods] or pods in standalone mode) not associated with a Replica Set (RS), Replication Controller (RC), Deployment, or Job (because these pods are not recreated).
```
Comment (mburke5678) [inline admin_guide/scheduling/descheduler.adoc]: Do we support these features: static/mirror pods; standalone mode? Nothing in our docs about it.

File: admin_guide/scheduling/descheduler.adoc [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+           "pods": 20
+         targetThresholds: <2>
+           "cpu" : 50
+           "memory": 50
+           "pods": 50
+         numberOfNodes: 3 <3>
+  "RemovePodsViolatingInterPodAntiAffinity": <1>
+     enabled: true
+----
+
```
Comment (mburke5678) [inline admin_guide/scheduling/descheduler.adoc]: Are the strategies intended to be in one file?

File: admin_guide/scheduling/descheduler.adoc [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
++
+----
+# oc create -f descheduler-job.yaml
+----
++
+For example:
++
+----
+# oc create -f descheduler-job.yaml
+----
```
Comment (mburke5678) [inline admin_guide/scheduling/descheduler.adoc]: Does this command make the descheduler "go"? Does it run once or keep going until you stop it?

File: admin_guide/scheduling/descheduler.adoc (hunk 2 above)
Comment (aveshagarwal) [reply admin_guide/scheduling/descheduler.adoc]: Yes, I mean descheduler policy has to be in one file.

File: admin_guide/scheduling/descheduler.adoc (hunk 3 above)
Comment (aveshagarwal) [reply admin_guide/scheduling/descheduler.adoc]: Yes Job will run descheduler in a pod.  Since it is going to run as Job, it will complete once descheduler exits. It wont run forever.

**Comment by aveshagarwal** [comment]
I went through the docs and it looks good overall to me.

**Comment by mburke5678** [comment]
@wjiangjay Please take a look.

**Comment by vikram-redhat** [comment]
[Document: Implement basic descheduler](https://trello.com/c/nulaT5Wp/670-document-implement-basic-descheduler)

File: admin_guide/scheduling/descheduler.adoc [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+# oc create sa <file-name>.yaml -n kube-system
+----
+
+. Bind the cluster role to the service account:
++
+----
+# oc create clusterrolebinding descheduler-cluster-role-binding \
+    --clusterrole=descheduler-cluster-role \
+    --serviceaccount=kube-system:descheduler-sa
```
Comment (wjiangjay) [inline admin_guide/scheduling/descheduler.adoc]: Since you hardcord the serviceaccount name but create sa prompt user to provide a name.

So please make it consistent.

**Comment by wjiangjay** [comment]
Could you provide a preview version do help review?

**Comment by mburke5678** [comment]
@wjiangjay 
Preview version with the change you requested:
http://file.rdu.redhat.com/mburke/trello-descheduler/trello-descheduler/admin_guide/scheduling/descheduler.html

File: admin_guide/scheduling/descheduler.adoc [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+== Overview
+
+Descheculing involves evicting pods based on xref:admin-guide-descheduler-policies[specific policies] so that the pods can be rescheduled
+onto more appropriate nodes. 
+
+Your cluster can benefit from descheduling and rescheduling already-running pods for various reasons:
+
+* A nodes is under- or over-utilized.
+* Pod and node affinity requirements, such as tains or labels, have changed and the original scheduling decisions are longer appropriate for cetain nodes.
```
Comment (wjiangjay) [inline admin_guide/scheduling/descheduler.adoc]: **taints** or labels?

Comment (wjiangjay) [reply admin_guide/scheduling/descheduler.adoc]: **certain** nodes?

File: admin_guide/scheduling/descheduler.adoc (hunk 1 above)
Comment (wjiangjay) [reply admin_guide/scheduling/descheduler.adoc]: Should include deploymentconfig here?

Comment (wjiangjay) [reply admin_guide/scheduling/descheduler.adoc]: And do OpenShift doc have static and mirror pods? if have, better link to OpenShift instead if kubernetes.

File: admin_guide/scheduling/descheduler.adoc [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+----
+
+<1> Set this value to `enabled: true` to use this policy. Set to `false` to disable this policy.
+
+[[admin-guide-descheduler-policies-low]]
+=== Creating a Low Node Utilization Policy
+
+The `LowNodeUtilization` strategy finds nodes that are underutilized and evicts pods from other nodes so that the evicted pods can be scheduled on these underutilized nodes. 
+
+The underutilization of nodes is determined by a configurable threshold, `thresholds`, for CPU, memory, or number of pods. If a node usage is below all these thresholds, the node is considered underutilized and the descheduler can evict pods from other nodes. Pods request resource requirements are considered when computing node resource utilization.
```
Comment (wjiangjay) [inline admin_guide/scheduling/descheduler.adoc]: number of pod?  should be percentage here @aveshagarwal

File: admin_guide/scheduling/descheduler.adoc (hunk 4 above)
Comment (wjiangjay) [reply admin_guide/scheduling/descheduler.adoc]: ignore this

**Comment by mburke5678** [comment]
@wjiangjay PTAL to see if I addressed your concerns.

**Comment by mburke5678** [comment]
@wjiangjay PTAL to see if I addressed your concerns.

**Comment by wjiangjay** [comment]
LGTM now.

**Comment by mburke5678** [comment]
[rev_history]
|xref:../admin_guide/scheduling/descheduler.adoc#admin-guide-descheduler[Scheduling -> Descheduler]
|New topic on the descheduler tool for proper distribution of pods for better cluster balancing.
%

**Comment by mburke5678** [comment]
@openshift/team-documentation PTAL

File: admin_guide/scheduling/descheduler.adoc [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+toc::[]
+
+== Overview
+
+Descheculing involves evicting pods based on xref:admin-guide-descheduler-policies[specific policies] so that the pods can be rescheduled
+onto more appropriate nodes. 
+
+Your cluster can benefit from descheduling and rescheduling already-running pods for various reasons:
+
+* A nodes is under- or over-utilized.
```
Comment (bfallonf) [inline admin_guide/scheduling/descheduler.adoc]: "A node is" or "Nodes are"

File: admin_guide/scheduling/descheduler.adoc [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+* Node failure requires pods to be moved.
+* New nodes are added to clusters.
+
+The descheduler does not schedule replacement of evicted pods. The xref:../../admin_guide/scheduling/scheduler.adoc#admin-guide-scheduler[scheduler]
+automatically performs this task for the evicted pods.
+
+[NOTE]
+====
+The descheduler job is considered a _critical pod_, which prevents the descheduler pod from being evicted by the descheduler. 
+You can make a pod a critical pod by adding the the `scheduler.alpha.kubernetes.io/critical-pod` annotation to the pod specification
```
Comment (bfallonf) [inline admin_guide/scheduling/descheduler.adoc]: I'm not sure this is the place to describe how to make a pod critical, especially if there's no context to why a pod should be critical. I'd suggest removing this, or putting in a little more context.

File: admin_guide/scheduling/descheduler.adoc [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+The descheduler does not schedule replacement of evicted pods. The xref:../../admin_guide/scheduling/scheduler.adoc#admin-guide-scheduler[scheduler]
+automatically performs this task for the evicted pods.
+
+[NOTE]
+====
+The descheduler job is considered a _critical pod_, which prevents the descheduler pod from being evicted by the descheduler. 
+You can make a pod a critical pod by adding the the `scheduler.alpha.kubernetes.io/critical-pod` annotation to the pod specification 
+====
+
+The descheduler job and descheduler pod are created in `kube-system` project.
```
Comment (bfallonf) [inline admin_guide/scheduling/descheduler.adoc]: "in _the_ `kube-system` project"

Comment (bfallonf) [reply admin_guide/scheduling/descheduler.adoc]: Also, is this project created by default? Do they have to have this project before running all this?

File: admin_guide/scheduling/descheduler.adoc [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+Best efforts pods are evicted before Burstable and Guaranteed pods.
+====
+
+To use the descheduler:
+
+. xref:admin-guide-descheduler-role[Create a role].
+. Define the descheduling behavior in a xref:admin-guide-descheduler-policies[policy file]. 
+. Create a xref:admin-guide-descheduler-policy[configuration map to reference the policy file].
+. Create the xref:admin-guide-descheduler-job[descheduler job configuration].
+. xref:admin-guide-descheduler-run[Run the descheduler job].
```
Comment (bfallonf) [inline admin_guide/scheduling/descheduler.adoc]: I was confused about the above. At first, I thought these were prerequisites, or links to outside content, but I now see we're describing the following sections. Can I suggest rewording the intro sentence?

"The following sections describe the process to configure the descheduler."

I'm not one for the list of sections in the topic (Minimalism! They're listed in other places!), but I know I might be alone in this. A reword of the intro might help my brain.

File: admin_guide/scheduling/descheduler.adoc [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+----
+# oc create clusterrolebinding descheduler-cluster-role-binding \
+    --clusterrole=descheduler-cluster-role \
+    --serviceaccount=kube-system:descheduler-sa
+----
+
+[[admin-guide-descheduler-policies]]
+== Creating Descheduler Policies
+
+The descheduler can be configured to remove pods from nodes that violate rules defined by _strategies_ in a YAML policy file. Include a path to the
```
Comment (bfallonf) [inline admin_guide/scheduling/descheduler.adoc]: s/The descheduler can be configured/You can configure the scheduler

File: admin_guide/scheduling/descheduler.adoc [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+           "memory": 20
+           "pods": 20
+         targetThresholds: <2>
+           "cpu" : 50
+           "memory": 50
+           "pods": 50
+         numberOfNodes: 3 <3>
+  "RemovePodsViolatingInterPodAntiAffinity": <1>
+     enabled: true
+----
```
Comment (bfallonf) [inline admin_guide/scheduling/descheduler.adoc]: Looks like there's footnotes that aren't attached to anything?

File: admin_guide/scheduling/descheduler.adoc [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+
+There are three default strategies that can be used with descheduler:
```
Comment (bfallonf) [inline admin_guide/scheduling/descheduler.adoc]: "with _the_ descheduler"

File: admin_guide/scheduling/descheduler.adoc [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+
+* Remove duplicate pods (`RemoveDuplicates`)
+* Move pods to underutilized nodes (`LowNodeUtilization`) 
+* Remove pods that violate anti-affinity rules (`RemovePodsViolatingInterPodAntiAffinity`). 
+
+The parameters associated with these strategies can be configured and strategies can be disabled as needed.
```
Comment (bfallonf) [inline admin_guide/scheduling/descheduler.adoc]: "You can configure and disable parameters associated with strategies as needed."

**Comment by mburke5678** [comment]
@aveshagarwal @wjiangjay @bfallonf 
I made some changes based on Brice's comments. Can I ask you to take a quick look again to make sure I haven't changed anything incorrectly. For ex. I think that `kube-system` is created by default. Just want to make sure. 
Also, I added a section to the "Managing Pods" topic on critical pods. We are supporting the `scheduler.alpha.kubernetes.io/critical-pod` annotation?

**Comment by mburke5678** [comment]
@aveshagarwal @wjiangjay 
I made some changes based on Brice's comments. Can I ask you to take a quick look again to make sure I haven't changed anything incorrectly. 
If I don't hear, I will merge and we can fix things later.

**Comment by mburke5678** [comment]
Per @aveshagarwal and @wjiangjay the Descheduler feature is now slated for 3.10 and needs to be removed from the 3.9 docs.

**Comment by mburke5678** [comment]
@aveshagarwal @wjiangjay 
Will this PR work for 3.10? Were there any changes to the feature when adding to 3.10?

**Comment by mburke5678** [comment]
@aveshagarwal @wjiangjay
Will this PR work for 3.10? Were there any changes to the feature when adding to 3.10?

**Comment by aveshagarwal** [comment]
@mburke5678 yes this should work for 3.10.
